src/backend/distributed/commands/foreign_server.c

```diff
@@ -0,0 +1,378 @@
+
+	Value *serverValue = linitial(stmt->objects);
+	ForeignServer *server = GetForeignServerByName(strVal(serverValue), false);
+
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+	/* unmark distributed server */
+	UnmarkObjectDistributed(&address);
+
+	const char *deparsedStmt = DeparseTreeNode((Node *) stmt);
+
+	/*
+	 * To prevent recursive propagation in mx architecture, we disable ddl
+	 * propagation before sending the command to workers.
+	 */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) deparsedStmt,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PostprocessCreateForeignServerStmt is called after a CREATE SERVER command has
+ * been executed by standard process utility.
+ */
+List *
+PostprocessCreateForeignServerStmt(Node *node, const char *queryString)
+{
+	bool missingOk = false;
+	ObjectAddress address = GetObjectAddressFromParseTree(node, missingOk);
+	EnsureDependenciesExistOnAllNodes(&address);
+
+	return NIL;
+}
+
+
+/*
+ * PostprocessAlterForeignServerOwnerStmt is called after a ALTER SERVER OWNER command
+ * has been executed by standard process utility.
+ */
+List *
+PostprocessAlterForeignServerOwnerStmt(Node *node, const char *queryString)
+{
+	bool missingOk = false;
+	ObjectAddress address = GetObjectAddressFromParseTree(node, missingOk);
+	EnsureDependenciesExistOnAllNodes(&address);
+
+	return NIL;
+}
+
+
+/*
+ * CreateForeignServerStmtObjectAddress finds the ObjectAddress for the server
+ * that is created by given CreateForeignServerStmt. If missingOk is false and if
+ * the server does not exist, then it errors out.
+ *
+ * Never returns NULL, but the objid in the address can be invalid if missingOk
+ * was set to true.
+ */
+ObjectAddress
+CreateForeignServerStmtObjectAddress(Node *node, bool missing_ok)
+{
+	CreateForeignServerStmt *stmt = castNode(CreateForeignServerStmt, node);
+
+	return GetObjectAddressByServerName(stmt->servername, missing_ok);
+}
+
+
+/*
+ * AlterForeignServerOwnerStmtObjectAddress finds the ObjectAddress for the server
+ * given in AlterOwnerStmt. If missingOk is false and if
+ * the server does not exist, then it errors out.
+ *
+ * Never returns NULL, but the objid in the address can be invalid if missingOk
+ * was set to true.
+ */
+ObjectAddress
+AlterForeignServerOwnerStmtObjectAddress(Node *node, bool missing_ok)
+{
+	AlterOwnerStmt *stmt = castNode(AlterOwnerStmt, node);
+	char *serverName = strVal(stmt->object);
+
+	return GetObjectAddressByServerName(serverName, missing_ok);
+}
+
+
+/*
+ * GetForeignServerCreateDDLCommand returns a list that includes the CREATE SERVER
+ * command that would recreate the given server on a new node.
+ */
+List *
+GetForeignServerCreateDDLCommand(Oid serverId)
+{
+	/* generate a statement for creation of the server in "if not exists" construct */
+	Node *stmt = RecreateForeignServerStmt(serverId);
```

Comment:
I really couldn't follow why we only have `RecreateForeignServerStmt` but say we don't have `ReCreateDropForeignServerStmt` or `ReCreateAlterForeignServerStmt`.

What makes `RecreateForeignServerStmt` special?

If you copy & pasted the logic from somewhere else, can you ask the original author why that is the case?

Reply:
This function is getting used to create the server on a worker when, say, we are doing a `citus_add_node`.
Here we generate a `CreateForeignServerStmt` with `if_not_exists = true`, then deparse it and send the command to workers.